File: extensions/drm-copilot/resources/scripts/dev_tools/pr_context/render_feature_excerpts.py
"""Feature-document excerpt helpers for PR context rendering."""

from __future__ import annotations

import re
from pathlib import Path
from typing import TYPE_CHECKING

from .models import FeatureDocExcerpt, section, truncate

if TYPE_CHECKING:
    from collections.abc import Iterable


def parse_section(markdown: str, heading: str) -> str:
    """Extract markdown content under a top-level `##` heading."""
    escaped = re.escape(heading)
    pattern = rf"^##\s+{escaped}\s*\r?\n(.*?)(?=^##\s+|\Z)"
    match = re.search(pattern, markdown, flags=re.MULTILINE | re.DOTALL)
    if not match:
        return ""
    return match.group(1).strip()


def _extract_issue_references(text: str) -> list[str]:
    """Extract normalized issue reference tokens from freeform text."""
    if not text:
        return []
    matches = re.findall(r"(?<!\w)#\d+|\b[A-Z][A-Z0-9]+-\d+\b", text)
    seen: set[str] = set()
    ordered: list[str] = []
    for item in matches:
        if item not in seen:
            seen.add(item)
            ordered.append(item)
    return ordered


def completed_plan_tasks(markdown: str, *, limit: int = 10) -> list[str]:
    """Return up to `limit` completed checklist items from markdown."""
    tasks: list[str] = []
    for line in markdown.splitlines():
        if re.search(r"\[x\]", line, flags=re.IGNORECASE):
            cleaned = re.sub(r"^[-*]\s*\[[xX]\]\s*", "", line).strip()
            tasks.append(cleaned)
        if len(tasks) >= limit:
            break
    return tasks


def directory_exists(path: Path) -> bool:
    """Return whether the provided path exists."""
    return path.exists()


def resolve_feature_dir(base_dir: Path, feature: str) -> Path | None:
    """Resolve feature directory by exact, strong-pattern, then weak match."""
    direct = base_dir / feature
    if directory_exists(direct):
        return direct

    if not directory_exists(base_dir):
        return None

    pattern = re.compile(rf"(?:^|[-_]){re.escape(feature)}(?:[-_]|$)")
    strong_matches: list[Path] = []
    weak_matches: list[Path] = []

    for candidate in sorted(base_dir.iterdir()):
        if not candidate.is_dir():
            continue
        name = candidate.name
        if pattern.search(name):
            strong_matches.append(candidate)
        elif feature in name:
            weak_matches.append(candidate)

    if strong_matches:
        return strong_matches[0]
    if weak_matches:
        return weak_matches[0]
    return None


def read_text_file(path: Path) -> str:
    """Read UTF-8 text if the path exists, otherwise return an empty string."""
    return path.read_text(encoding="utf-8") if path.exists() else ""


def extract_features_from_paths(changed_files: Iterable[str]) -> set[str]:
    """Extract feature directory names from `docs/features/active/**` paths."""
    features: set[str] = set()
    for raw in changed_files:
        parts = Path(raw).parts
        if (
            len(parts) >= 4
            and parts[0] == "docs"
            and parts[1] == "features"
            and parts[2] == "active"
        ):
            features.add(parts[3])
    return features


def extract_spec_parts(spec_text: str) -> list[str]:
    """Extract high-value sections from a feature spec document."""
    spec_parts: list[str] = []
    for heading in (
        "Context",
        "Root Cause",
        "Root Cause/Problem",
        "Problem",
        "Proposed Fix",
        "Acceptance Criteria",
        "Constraints & Risks",
        "Behavior",
        "Overview",
    ):
        section_text = parse_section(spec_text, heading)
        if section_text:
            spec_parts.append(f"{heading}: {truncate(section_text)}")
    return spec_parts


def extract_plan_sections(plan_text: str) -> tuple[str, str]:
    """Extract completed tasks and verification notes from a plan document."""
    plan_tasks = completed_plan_tasks(plan_text)
    plan_section = "\n".join(f"- {task}" for task in plan_tasks) if plan_tasks else ""
    # Keep fallback order aligned with feature_docs for consistent semantics.
    test_plan_section = parse_section(plan_text, "Verification")
    if not test_plan_section:
        test_plan_section = parse_section(plan_text, "Test Plan")
    verification_block = (
        "Plan verification notes:\n" + truncate(test_plan_section)
        if test_plan_section
        else ""
    )
    return plan_section, verification_block


def extract_story_parts(user_story_text: str, promoted_story_text: str) -> list[str]:
    """Extract story statement/problem snippets from user-story content."""
    story_parts: list[str] = []
    story_statements = parse_section(user_story_text, "Story Statement")
    if story_statements:
        story_lines = [
            line.strip("- ") for line in story_statements.splitlines() if line.strip()
        ]
        if story_lines:
            story_parts.append(
                "Story Statement:\n" + "\n".join(f"- {line}" for line in story_lines)
            )
    problem_section = parse_section(user_story_text, "Problem / Why")
    if problem_section:
        story_parts.append("Problem / Why:\n" + truncate(problem_section))
    if not story_parts and promoted_story_text:
        promoted_problem = parse_section(promoted_story_text, "Problem / Why")
        if not promoted_problem:
            promoted_problem = parse_section(promoted_story_text, "Summary")
        if promoted_problem:
            story_parts.append("Problem / Why:\n" + truncate(promoted_problem))
    return story_parts


def build_excerpt_text(
    feature: str,
    story_parts: list[str],
    spec_parts: list[str],
    plan_section: str,
    verification_block: str,
) -> str:
    """Build a formatted feature excerpt block from collected sections."""
    lines: list[str] = [section(f"Feature doc: {feature}")]
    if story_parts:
        lines.append("User story excerpts:\n" + "\n\n".join(story_parts))
    if spec_parts:
        lines.append("Spec excerpts:\n" + "\n\n".join(spec_parts))
    if plan_section:
        lines.append("Plan completed tasks:\n" + plan_section)
    if verification_block:
        lines.append(verification_block)
    if len(lines) == 1:
        lines.append("(no spec/plan/user-story excerpts found)")
    return "\n".join(lines)


def gather_feature_excerpts(
    root: Path, changed_files: Iterable[str]
) -> list[FeatureDocExcerpt]:
    """Gather feature excerpt payloads for changed active-feature files."""
    features = extract_features_from_paths(changed_files)
    excerpts: list[FeatureDocExcerpt] = []
    base_dir = root / "docs" / "features" / "active"
    promoted_dir = root / "docs" / "features" / "potential" / "promoted"

    for feature in sorted(features):
        feature_dir = resolve_feature_dir(base_dir, feature)
        promoted_feature_dir = resolve_feature_dir(promoted_dir, feature)
        if feature_dir is None and promoted_feature_dir is None:
            continue

        active_dir = feature_dir or promoted_feature_dir
        if active_dir is None:
            continue

        spec_path = active_dir / "spec.md"
        plan_path = active_dir / "plan.md"
        user_story_path: Path = active_dir / "user-story.md"
        promoted_story_path = (
            promoted_feature_dir / "user-story.md"
            if promoted_feature_dir is not None
            else None
        )
        promoted_story_text = (
            read_text_file(promoted_story_path) if promoted_story_path else ""
        )
        if promoted_story_path is not None and not user_story_path.exists():
            user_story_path = promoted_story_path

        user_story_text = read_text_file(user_story_path)
        if (
            not user_story_text
            and promoted_story_text
            and promoted_story_path is not None
        ):
            user_story_text = promoted_story_text
            user_story_path = promoted_story_path

        spec_text = read_text_file(spec_path)
        plan_text = read_text_file(plan_path)

        spec_parts = extract_spec_parts(spec_text)
        plan_section, verification_block = extract_plan_sections(plan_text)
        story_parts = extract_story_parts(user_story_text, promoted_story_text)
        excerpt_text = build_excerpt_text(
            feature, story_parts, spec_parts, plan_section, verification_block
        )

        context_files = [
            str(path.relative_to(root))
            for path in (spec_path, plan_path, user_story_path)
            if path and path.exists()
        ]
        issue_refs = _extract_issue_references(
            "\n".join([spec_text, plan_text, user_story_text])
        )
        excerpts.append(
            FeatureDocExcerpt(
                feature=feature,
                excerpt=excerpt_text,
                issue_refs=issue_refs,
                context_files=context_files,
            )
        )

    return excerpts

File: extensions/drm-copilot/src/lib/pr-context/render-feature-excerpts.ts
/**
 * Feature-document excerpt helpers for PR context rendering (render variant).
 *
 * Purpose:
 *     Port of `dev_tools/pr_context/render_feature_excerpts.py`. This is the
 *     render-module variant: it produces the smaller `contextFiles` set
 *     (spec/plan/user-story only) and carries no readiness or primary-issue
 *     fields, distinct from the collector `feature-docs.ts` variant.
 *
 * Responsibilities:
 *     - `parseSection`, `completedPlanTasks`, `extractIssueReferences`.
 *     - `directoryExists`, `resolveFeatureDir`, `readTextFile`.
 *     - `extractFeaturesFromPaths`, `extractSpecParts`, `extractPlanSections`,
 *       `extractStoryParts`, `buildExcerptText`, and `gatherFeatureExcerpts`.
 *     All filesystem access flows through the injected {@link FileSystem}.
 */

import { type FileSystem, toPosixPath } from "../file-system";
import {
  type FeatureDocExcerpt,
  section,
  splitLines,
  truncate,
} from "./models";

/** Spec headings extracted into the excerpt, in fixed order. */
const SPEC_HEADINGS = [
  "Context",
  "Root Cause",
  "Root Cause/Problem",
  "Problem",
  "Proposed Fix",
  "Acceptance Criteria",
  "Constraints & Risks",
  "Behavior",
  "Overview",
] as const;

/**
 * Extract markdown content under a top-level `##` heading.
 *
 * Mirrors Python `parse_section`.
 *
 * @param markdown Markdown content.
 * @param heading Heading to match.
 * @returns The trimmed section body, or `""`.
 */
export function parseSection(markdown: string, heading: string): string {
  const escaped = escapeRegExp(heading);
  const pattern = new RegExp(
    `^##\\s+${escaped}\\s*\\r?\\n([\\s\\S]*?)(?=^##\\s+|$(?![\\s\\S]))`,
    "m",
  );
  const match = pattern.exec(markdown);
  if (!match) {
    return "";
  }
  return match[1]!.trim();
}

/**
 * Extract normalized issue reference tokens from freeform text.
 *
 * Mirrors Python `_extract_issue_references`.
 *
 * @param text Source text.
 * @returns Ordered, deduplicated reference tokens.
 */
export function extractIssueReferences(text: string): string[] {
  if (!text) {
    return [];
  }
  const matches = text.match(/(?<!\w)#\d+|\b[A-Z][A-Z0-9]+-\d+\b/gu) ?? [];
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const item of matches) {
    if (!seen.has(item)) {
      seen.add(item);
      ordered.push(item);
    }
  }
  return ordered;
}

/**
 * Return up to `limit` completed checklist items from markdown.
 *
 * Mirrors Python `completed_plan_tasks`.
 *
 * @param markdown Markdown content.
 * @param limit Maximum tasks returned (default 10).
 * @returns The completed-task texts.
 */
export function completedPlanTasks(markdown: string, limit = 10): string[] {
  const tasks: string[] = [];
  for (const line of splitLines(markdown)) {
    if (/\[x\]/iu.test(line)) {
      tasks.push(line.replace(/^[-*]\s*\[[xX]\]\s*/u, "").trim());
    }
    if (tasks.length >= limit) {
      break;
    }
  }
  return tasks;
}

/**
 * Return whether the provided path exists.
 *
 * Mirrors Python `directory_exists`.
 *
 * @param fs Injected filesystem.
 * @param path Path to test.
 * @returns True when the path exists.
 */
export function directoryExists(fs: FileSystem, path: string): boolean {
  return fs.exists(path);
}

/**
 * Resolve a feature directory by exact, strong-pattern, then weak match.
 *
 * Mirrors Python `resolve_feature_dir`: exact child; otherwise — when the base
 * directory exists — iterate sorted children classifying strong/weak matches.
 *
 * @param fs Injected filesystem.
 * @param baseDir Base directory.
 * @param feature Feature identifier.
 * @returns The resolved directory path, or `null`.
 */
export function resolveFeatureDir(
  fs: FileSystem,
  baseDir: string,
  feature: string,
): string | null {
  const direct = `${baseDir}/${feature}`;
  if (directoryExists(fs, direct)) {
    return direct;
  }

  if (!directoryExists(fs, baseDir)) {
    return null;
  }

  const pattern = new RegExp(
    `(?:^|[-_])${escapeRegExp(feature)}(?:[-_]|$)`,
    "u",
  );
  const strongMatches: string[] = [];
  const weakMatches: string[] = [];

  // Iterate sorted children, classifying directories by match strength.
  for (const name of fs.listDirectory(baseDir)) {
    const candidate = `${baseDir}/${name}`;
    if (!fs.isDirectory(candidate)) {
      continue;
    }
    if (pattern.test(name)) {
      strongMatches.push(candidate);
    } else if (name.includes(feature)) {
      weakMatches.push(candidate);
    }
  }

  if (strongMatches.length > 0) {
    return strongMatches[0]!;
  }
  if (weakMatches.length > 0) {
    return weakMatches[0]!;
  }
  return null;
}

/**
 * Read UTF-8 text if the path exists, otherwise return an empty string.
 *
 * Mirrors Python `read_text_file`.
 *
 * @param fs Injected filesystem.
 * @param path File path.
 * @returns The file content, or `""`.
 */
export function readTextFile(fs: FileSystem, path: string): string {
  return fs.exists(path) ? fs.readTextFile(path) : "";
}

/**
 * Extract feature directory names from `docs/features/active/**` paths.
 *
 * Mirrors Python `extract_features_from_paths`.
 *
 * @param changedFiles Repo-relative changed file paths.
 * @returns The set of active-feature directory names.
 */
export function extractFeaturesFromPaths(
  changedFiles: Iterable<string>,
): Set<string> {
  const features = new Set<string>();
  for (const raw of changedFiles) {
    const parts = toPosixPath(raw).split("/");
    if (
      parts.length >= 4 &&
      parts[0] === "docs" &&
      parts[1] === "features" &&
      parts[2] === "active"
    ) {
      features.add(parts[3]!);
    }
  }
  return features;
}

/**
 * Extract high-value sections from a feature spec document.
 *
 * Mirrors Python `extract_spec_parts`.
 *
 * @param specText Spec markdown.
 * @returns The spec excerpt parts.
 */
export function extractSpecParts(specText: string): string[] {
  const specParts: string[] = [];
  for (const heading of SPEC_HEADINGS) {
    const sectionText = parseSection(specText, heading);
    if (sectionText) {
      specParts.push(`${heading}: ${truncate(sectionText)}`);
    }
  }
  return specParts;
}

/**
 * Extract completed tasks and verification notes from a plan document.
 *
 * Mirrors Python `extract_plan_sections`: `Verification` then `Test Plan`
 * fallback for the verification block.
 *
 * @param planText Plan markdown.
 * @returns A tuple of `[planSection, verificationBlock]`.
 */
export function extractPlanSections(planText: string): [string, string] {
  const planTasks = completedPlanTasks(planText);
  const planSection =
    planTasks.length > 0 ? planTasks.map((task) => `- ${task}`).join("\n") : "";
  let testPlanSection = parseSection(planText, "Verification");
  if (!testPlanSection) {
    testPlanSection = parseSection(planText, "Test Plan");
  }
  const verificationBlock = testPlanSection
    ? "Plan verification notes:\n" + truncate(testPlanSection)
    : "";
  return [planSection, verificationBlock];
}

/**
 * Extract story statement/problem snippets from user-story content.
 *
 * Mirrors Python `extract_story_parts`.
 *
 * @param userStoryText Active user-story markdown.
 * @param promotedStoryText Promoted-folder user-story markdown fallback.
 * @returns The story excerpt parts.
 */
export function extractStoryParts(
  userStoryText: string,
  promotedStoryText: string,
): string[] {
  const storyParts: string[] = [];
  const storyStatements = parseSection(userStoryText, "Story Statement");
  if (storyStatements) {
    const storyLines = splitLines(storyStatements)
      .filter((line) => line.trim())
      .map((line) => line.replace(/^[- ]+|[- ]+$/gu, ""));
    if (storyLines.length > 0) {
      storyParts.push(
        "Story Statement:\n" + storyLines.map((line) => `- ${line}`).join("\n"),
      );
    }
  }
  const problemSection = parseSection(userStoryText, "Problem / Why");
  if (problemSection) {
    storyParts.push("Problem / Why:\n" + truncate(problemSection));
  }
  if (storyParts.length === 0 && promotedStoryText) {
    let promotedProblem = parseSection(promotedStoryText, "Problem / Why");
    if (!promotedProblem) {
      promotedProblem = parseSection(promotedStoryText, "Summary");
    }
    if (promotedProblem) {
      storyParts.push("Problem / Why:\n" + truncate(promotedProblem));
    }
  }
  return storyParts;
}

/**
 * Build a formatted feature excerpt block from collected sections.
 *
 * Mirrors Python `build_excerpt_text`.
 *
 * @param feature Feature identifier.
 * @param storyParts Story excerpt parts.
 * @param specParts Spec excerpt parts.
 * @param planSection Completed-tasks block.
 * @param verificationBlock Verification block.
 * @returns The excerpt text.
 */
export function buildExcerptText(
  feature: string,
  storyParts: string[],
  specParts: string[],
  planSection: string,
  verificationBlock: string,
): string {
  const lines: string[] = [section(`Feature doc: ${feature}`)];
  if (storyParts.length > 0) {
    lines.push("User story excerpts:\n" + storyParts.join("\n\n"));
  }
  if (specParts.length > 0) {
    lines.push("Spec excerpts:\n" + specParts.join("\n\n"));
  }
  if (planSection) {
    lines.push("Plan completed tasks:\n" + planSection);
  }
  if (verificationBlock) {
    lines.push(verificationBlock);
  }
  if (lines.length === 1) {
    lines.push("(no spec/plan/user-story excerpts found)");
  }
  return lines.join("\n");
}

/**
 * Gather feature excerpt payloads for changed active-feature files (render
 * variant).
 *
 * Mirrors Python `gather_feature_excerpts` in `render_feature_excerpts.py`:
 * smaller `contextFiles` (spec/plan/user-story only), no readiness or
 * primary-issue fields.
 *
 * @param fs Injected filesystem.
 * @param root Repository root.
 * @param changedFiles Repo-relative changed file paths.
 * @returns One excerpt per resolved feature, sorted by feature name.
 */
export function gatherFeatureExcerpts(
  fs: FileSystem,
  root: string,
  changedFiles: Iterable<string>,
): FeatureDocExcerpt[] {
  const normalizedRoot = toPosixPath(root).replace(/\/+$/u, "");
  const features = extractFeaturesFromPaths(changedFiles);
  const excerpts: FeatureDocExcerpt[] = [];
  const baseDir = `${normalizedRoot}/docs/features/active`;
  const promotedDir = `${normalizedRoot}/docs/features/potential/promoted`;

  for (const feature of [...features].sort(compareCodePoint)) {
    const featureDir = resolveFeatureDir(fs, baseDir, feature);
    const promotedFeatureDir = resolveFeatureDir(fs, promotedDir, feature);
    if (featureDir === null && promotedFeatureDir === null) {
      continue;
    }
    const activeDir = featureDir ?? promotedFeatureDir;
    if (activeDir === null) {
      continue;
    }

    const specPath = `${activeDir}/spec.md`;
    const planPath = `${activeDir}/plan.md`;
    let userStoryPath = `${activeDir}/user-story.md`;
    const promotedStoryPath =
      promotedFeatureDir !== null
        ? `${promotedFeatureDir}/user-story.md`
        : null;
    const promotedStoryText =
      promotedStoryPath !== null ? readTextFile(fs, promotedStoryPath) : "";
    if (promotedStoryPath !== null && !fs.exists(userStoryPath)) {
      userStoryPath = promotedStoryPath;
    }

    let userStoryText = readTextFile(fs, userStoryPath);
    if (!userStoryText && promotedStoryText && promotedStoryPath !== null) {
      userStoryText = promotedStoryText;
      userStoryPath = promotedStoryPath;
    }

    const specText = readTextFile(fs, specPath);
    const planText = readTextFile(fs, planPath);

    const specParts = extractSpecParts(specText);
    const [planSection, verificationBlock] = extractPlanSections(planText);
    const storyParts = extractStoryParts(userStoryText, promotedStoryText);
    const excerptText = buildExcerptText(
      feature,
      storyParts,
      specParts,
      planSection,
      verificationBlock,
    );

    // Smaller context-files set: existing spec/plan/user-story only, in order.
    const contextFiles: string[] = [];
    for (const path of [specPath, planPath, userStoryPath]) {
      if (fs.exists(path)) {
        contextFiles.push(relativeToPosix(normalizedRoot, path));
      }
    }
    const issueRefs = extractIssueReferences(
      [specText, planText, userStoryText].join("\n"),
    );
    excerpts.push({
      feature,
      excerpt: excerptText,
      issueRefs,
      contextFiles,
      primaryIssueRef: null,
      readinessSignal: null,
    });
  }

  return excerpts;
}

/** Compute a repo-relative POSIX path for a path under `root`. */
function relativeToPosix(root: string, path: string): string {
  const normalized = toPosixPath(path);
  if (normalized.startsWith(`${root}/`)) {
    return normalized.slice(root.length + 1);
  }
  return normalized.replace(/^\/+/u, "");
}

/** Compare two strings by Unicode code point (Python `sorted` semantics). */
function compareCodePoint(left: string, right: string): number {
  if (left < right) {
    return -1;
  }
  if (left > right) {
    return 1;
  }
  return 0;
}

/** Escape regex metacharacters for a dynamic heading pattern. */
function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&");
}
